### ISSUE TITLE
No access to default arguments in Mirrors

### ISSUE THREAD

**Issue by dwijnand** [lampepfl/dotty-feature-requests#162]
## Minimized code

```scala
case class WithDefault(id: Int, a: String = "a", b: Option[String] = Some("b"))
```

## Output

```scala
scala> import scala.compiletime._, scala.deriving._

scala> val m = summon[Mirror.Of[WithDefault]]
val m:
  (
    deriving.Mirror{
      MirroredType = WithDefault; MirroredMonoType = WithDefault;
        MirroredElemTypes <: Tuple
    }
   &
    scala.deriving.Mirror.Product{
      MirroredMonoType = WithDefault; MirroredType = WithDefault;
        MirroredLabel = "WithDefault"
    }
  ){
    MirroredElemTypes = (Int, String, Option[String]);
      MirroredElemLabels = ("id", "a", "b")
  } = WithDefault
```

## Expectation

I would expect a way to access to default arguments, such that it may be used in a derived typeclass instance.

**Comment by SethTisue** [comment]
> We should probably look at it only once we have revamped default parameters in Tasty

(Has that happened, by any chance?)

**Comment by bishabosha** [comment]
Not yet, it's still the same encoding with name mangled methods

**Comment by bishabosha** [comment]
no one is working on this as far as I know. Do you have a specific API in mind?

**Comment by godenji** [comment]
> Do you have a specific API in mind?

yes, Play JSON, but I assume any Scala 3 json library that attempts to extract case class defaut arguments at compile time is affected.

**Comment by odersky** [comment]
Changing the default parameter scheme is a very large undertaking. I would guess on the order of 3-6pm. I won't have the time to do it personally. I am not sure how we will be able to fund/resource this. Suggestions are welcome.

**Comment by Lasering** [comment]
While this is not implemented see this solution using macros: https://stackoverflow.com/questions/68421043/type-class-derivation-accessing-default-values

**Comment by adampauls** [comment]
I'm not sure if this makes anything easier, but a slightly more limited option would be to expose a fromNamedProduct method on mirrors that takes a map from names to values. You wouldn't be able to access the default values directly, but you could at least construct an instance where the default are filled in for you. Might that avoid the need for any major refactors?

**Comment by odersky** [comment]
The problem is that the "default values" are hidden in default getter methods. And, if the default depends on previous parameters it will be a function over those parameter values instead of being a simple value. 

For the moment, I think the only solution would be the one @Lasering describes.

**Comment by adampauls** [comment]
Thanks, that much I understood. I don't follow why a compiler-generated `fromNamedProduct` method couldn't assemble the right default function invocations (taking into account dependencies on other supplied arguments) on the callers behalf, but of course I'll defer. Maybe I'm wrong in thinking that `fromProduct` implementations are compiler-generated.

**Comment by odersky** [comment]
Yes, I see. `fromNamedProduct` could possibly work. But it would be pretty heavy on the code generation side. We were very careful so far not to overburden the code that gets generated for every case class. So far, that's just `fromProduct`, which is pretty lightweight. Adding `fromNamedProduct` to all case classes could blow our budget, but it would not be an issue if `fromNamedProduct` was only generated when certain traits were derived.

**Comment by Lasering** [comment]
I think @adampauls idea was to (my emphasis):
> expose a fromNamedProduct method **on mirrors**

For `Mirror.Product` there is already a `fromProduct` method. Having the code generation only happen on Mirrors shouldn't blow the budget for case class code generation. Unless those two are considered to be in the same bucket.

**Comment by bishabosha** [comment]
> I think @adampauls idea was to (my emphasis):
> 
> > expose a fromNamedProduct method **on mirrors**
> 
> Having the code generation only happen on Mirrors shouldn't blow the budget for case class code generation. Unless those two are considered to be in the same bucket.

The companion class of every case class compiled by scala 3 is also the Mirror
